### PR TITLE
Fetch comments data with JSON formatting

### PR DESCRIPTION
**Pull request by difanyang**
Added a JSON-formatted ArrayList of Strings to the page.

### CODE REVIEW

File: portfolio/src/main/webapp/script.js [hunk 1]
```diff
@@ -27,8 +27,19 @@ function addRandomFacts() {
   factContainer.innerText = fact;
 }
 
-/** Fetches the content in DataServlet and display on Home page */
+/** Fetches comments from the servers and adds them to the DOM. */
 function getData() {
-  fetch('/data').then(response => response.text()).then((data) => 
-  {document.getElementById("data-container").innerText = data;});
+  fetch('/data').then(response => response.json()).then((data) => 
+  {
```
Comment (floatfeather) [inline portfolio/src/main/webapp/script.js:33]: Please do not add a new line for '{'.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 2]
```diff
@@ -23,10 +25,18 @@
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
+  static ArrayList<String> stringList = new ArrayList<String>();
```
Comment (zixiyan) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:28]: static private final?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 3]
```diff
@@ -23,10 +25,18 @@
...
+    stringList.add("Nice to meet you!");
+    stringList.add("Have a nice day!");
+    stringList.add("Stay safe and healthy!");
+  }
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("Hello Difan Yang!");
+    Gson gson = new Gson();
```
Comment (zixiyan) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:37]: make it static final?
```
static final Gson GSON = new Gson();
```

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 4]
```diff
@@ -23,10 +25,18 @@
...
+    String json = gson.toJson(stringList);
```
Comment (zixiyan) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:38]: try inline?
```
response.getWriter().println(GSON.toJson(...))
```

**Comment by difanyang** [comment]
These issues have been resolved in PR #7

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 5]
```diff
@@ -25,18 +28,37 @@
...
 public class DataServlet extends HttpServlet {
-  static ArrayList<String> stringList = new ArrayList<String>();
-  static {
-    stringList.add("Nice to meet you!");
-    stringList.add("Have a nice day!");
-    stringList.add("Stay safe and healthy!");
-  }
+  /** Initially empty ArrayList. Later user comments will be stored as elements */
+  private static ArrayList<String> stringList = new ArrayList<String>();
+  static final Gson GSON = new Gson();
```
Comment (zixiyan) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:33]: static private final?

**Comment by difanyang** [comment]
Sad story: I accidentally erased a number of commits with a single force-push and searched for hours the way to restore data without success.

File: portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+import javax.servlet.http.HttpServletResponse;
+
+/** Servlet responsible for deleting comments. */
+@WebServlet("/delete-data")
+public class DeleteDataServlet extends HttpServlet {
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Query query = new Query("Comment");
```
Comment (zixiyan) [inline portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java:24]: consider to use https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/datastore/Query.html#setkeysonly ?
So you could save some query time.

File: portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    PreparedQuery results = datastore.prepare(query);
+
+    for (Entity entity : results.asIterable()) {
+      Key key = entity.getKey();
+      datastore.delete(key);
```
Comment (zixiyan) [inline portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java:29]: consider inline?
```
datastore.delete(entity.getKey());
```

File: portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** Servlet responsible for deleting comments. */
+@WebServlet("/delete-data")
+public class DeleteDataServlet extends HttpServlet {
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
```
Comment (zixiyan) [inline portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java:23]: consider to make it private static final?
just like what you did above?

**Comment by zixiyan** [comment]
Is this your latest PR?

File: portfolio/src/main/webapp/script.js [hunk 9]
```diff
@@ -27,8 +27,24 @@ function addRandomFacts() {
...
+
+/** Creates an <li> element containing text. */
+function createListElement(text) {
+  const liElement = document.createElement('li');
+  liElement.innerText = text;
+  return liElement;
+}
+
+function deleteData(){
+  fetch('/delete-data');
```
Comment (zixiyan) [inline portfolio/src/main/webapp/script.js:49]: If you want to POST request, it seems that you need to specify explicitly.
```
fetch('/delete-data', {method: 'post'})
```
please refer to https://googlechrome.github.io/samples/fetch-api/fetch-post.html for details.

**Comment by difanyang** [comment]
> Is this your latest PR?

Yes.

File: portfolio/src/main/webapp/index.html [hunk 10]
```diff
@@ -16,6 +16,21 @@
...
+    </form>
+    <form action="/data" method="GET">
+      <label for="numComment">Number of comments to display:</label>
+      <input type="number" name="numComment" value="1">
+      </br></br>
+      <input type="submit" />
+      </br></br>
+    </form>
+    <button onclick="deleteData()">Delete all comments</button>
+    <ul id="data-container"></ul>
```
Comment (zixiyan) [inline portfolio/src/main/webapp/index.html:34]: consider to rename it to "comments-container"

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 11]
```diff
@@ -14,19 +14,64 @@
...
+
+    response.setContentType("application/json;");
+    response.getWriter().println(GSON.toJson(stringList));
+    response.sendRedirect("index.html");
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Get the input from the form
+    String text = getParameterWithDefault(request, "comment", "");
```
Comment (zixiyan) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:58]: consider to skip empty comment?

**Comment by difanyang** [comment]
Just realized I did not need to create and merge that PR since the branch had already been merged.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 12]
```diff
@@ -14,19 +14,64 @@
...
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
-/** Servlet that returns some example content. TODO: modify this file to handle comments data */
+/** Servlet that returns user comments. TODO: modify this file to handle comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
+  private static final Gson GSON = new Gson();
+  private static final DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
```
Comment (zixiyan) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:36]: per styleguide, static variables like ```datastore``` should be renamed to ```DATASTORE```

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 11 above)
Comment (zixiyan) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:57]: Please use empty to provide better readability.
```
!text.empty()
```

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 13]
```diff
@@ -14,19 +14,64 @@
...
-    response.setContentType("text/html;");
-    response.getWriter().println("Hello Difan Yang!");
+    Query query = new Query("Comment");
+    PreparedQuery results = datastore.prepare(query);
+    ArrayList<String> stringList = new ArrayList<String>();
+    int numComment = Integer.parseInt(getParameterWithDefault(request, "numComment", "0"));
+
+    for (final Entity entity : results.asIterable(FetchOptions.Builder.withLimit(numComment))) {
+      String comment = (String) entity.getProperty("content");
+      stringList.add(comment);
```
Comment (zixiyan) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:47]: consider inlining.
```
stringList.add(entity.getProperty("content"));
```

File: portfolio/src/main/webapp/script.js [hunk 14]
```diff
@@ -27,8 +27,22 @@ function addRandomFacts() {
   factContainer.innerText = fact;
 }
 
-/** Fetches the content in DataServlet and display on Home page */
+/** Fetches comments from the servers and adds them to the DOM. */
 function getData() {
-  fetch('/data').then(response => response.text()).then((data) => 
-  {document.getElementById("data-container").innerText = data;});
+  fetch('/data?numComment='+document.getElementById("numComment").value).then(response => response.json()).then((data) => {
```
Comment (zixiyan) [inline portfolio/src/main/webapp/script.js:32]: exceed 80 characters?

File: portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** Servlet responsible for deleting comments. */
+@WebServlet("/delete-data")
+public class DeleteDataServlet extends HttpServlet {
+  private static final DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
```
Comment (zixiyan) [inline portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java:20]: ditto.

File: portfolio/src/main/webapp/script.js [hunk 16]
```diff
@@ -27,8 +27,23 @@ function addRandomFacts() {
   factContainer.innerText = fact;
 }
 
-/** Fetches the content in DataServlet and display on Home page */
+/** Fetches comments from the servers and adds them to the DOM. */
 function getData() {
-  fetch('/data').then(response => response.text()).then((data) => 
-  {document.getElementById("data-container").innerText = data;});
+  fetch('/data?numComment='+document.getElementById("numComment").value).then(response
+   => response.json()).then((data) => {
```
Comment (zixiyan) [inline portfolio/src/main/webapp/script.js:33]: I would suggest use 4 spaces for breaking a line. And I would suggest a new line from ```.then```
```
fetch('/data?numComment='+document.getElementById("numComment").value).
    then(response=> response.json()).then((data) => {
  ...
}
```

**Comment by zixiyan** [comment]
I guess the two bugs in your PR have been fixed?

**Comment by difanyang** [comment]
> I guess the two bugs in your PR have been fixed?

Yes (finally...)

**Pull request merged**